bluemira/equilibria/optimisation/harmonics/toroidal_harmonics_approx_functions.py

```diff
@@ -422,224 +527,662 @@ def toroidal_harmonic_approximate_psi(
-        max_degree = len(th_params.th_coil_names) - 1
-
     # Get coil positions and currents from equilibrium
     currents = np.array([eq.coilset[name].current for name in th_params.th_coil_names])
 
+    cos_empty = len(cos_m_chosen) == 0
+    sin_empty = len(sin_m_chosen) == 0
+
     # Initialise psi and A arrays
     approx_coilset_psi = np.zeros_like(th_params.R)
     A = np.zeros_like(th_params.R)
     # Useful combination
     Delta = np.cosh(th_params.tau) - np.cos(th_params.sigma)  # noqa: N806
     # Get sigma values for the grid
-    sigma_mult_degree = [m * th_params.sigma for m in range(max_degree)]
+    sigma_mult_mode_cos = [m * th_params.sigma for m in cos_m_chosen]
+    sigma_mult_mode_sin = [m * th_params.sigma for m in sin_m_chosen]
+
+    factorial_m_cos = np.array([factorial(m) for m in cos_m_chosen])
+    factorial_m_sin = np.array([factorial(m) for m in sin_m_chosen])
 
-    epsilon = 2 * np.ones(max_degree)
-    epsilon[0] = 1
-    factorial_m = np.array([factorial(m) for m in range(max_degree)])
-    degrees = np.arange(0, max_degree)[:, None, None]
     # TH coefficient matrix
-    Am_cos, Am_sin = coil_toroidal_harmonic_amplitude_matrix(  # noqa: N806
-        input_coils=eq.coilset,
-        th_params=th_params,
-        max_degree=max_degree,
+    Am_cos_current_function, Am_sin_current_function = (  # noqa: N806
+        coil_toroidal_harmonic_amplitude_matrix(
+            input_coils=eq.coilset,
+            th_params=th_params,
+            cos_m_chosen=cos_m_chosen,
+            sin_m_chosen=sin_m_chosen,
+        )
     )
 
-    Am_cos_sin = np.einsum(  # noqa: N806
-        "ij, ikl -> ijkl", Am_cos, np.cos(sigma_mult_degree)
-    ) + np.einsum("ij, ikl -> ijkl", Am_sin, np.sin(sigma_mult_degree))
-    A = np.sqrt(2 / np.pi) * (
-        np.einsum(
-            "ijkl, i, i, kl, ikl, j -> kl",
-            Am_cos_sin,
-            epsilon,
-            factorial_m,
-            np.sqrt(Delta),
-            legendre_q(degrees - 1 / 2, 1, np.cosh(th_params.tau), n_max=30),
-            currents,
+    if cos_empty:
+        A_cos = 0  # noqa: N806
+    else:
+        epsilon = 2 * np.ones(len(cos_m_chosen))
+        epsilon[0] = 1
+        Am_cos_matrix = (  # noqa: N806
+            0
+            if cos_empty
+            else np.einsum(
+                "ij, ikl, i, ikl -> ijkl",
+                Am_cos_current_function,
+                np.cos(sigma_mult_mode_cos),
+                factorial_m_cos,
+                legendre_q(
+                    cos_m_chosen[:, None, None] - 1 / 2,
+                    1,
+                    np.cosh(th_params.tau),
+                    n_max=30,
+                ),
+            )
         )
-    )
+        A_cos = np.sqrt(2 / np.pi) * (  # noqa: N806
+            np.einsum(
+                "ijkl, i, kl, j -> kl",
+                Am_cos_matrix,
+                epsilon,
+                np.sqrt(Delta),
+                currents,
+            )
+        )
```

Comment:
this could all be a separate function maybe? Same for the sin. I'm probably mistaken here but they nearly look structurally the same apart from the choice of a few arguments. Meaning it could be one function and then here it will be two lines:

```python
A_cos = func(...)
A_sin = func(...)
```